File: frontend/src/lib/gst.ts
// Client-side GST preview, mirroring the Go engine (internal/service/gst.go) so
// the cashier sees live totals as they build a bill. The AUTHORITATIVE totals
// are always recomputed by the backend on save — this is display only.
//
// All amounts are in paise (integers), matching money.Money.

export interface PreviewLine {
  unitPrice: number;
  quantity: number;
  discount: number;
  gstRate: number;
}

export interface PreviewTotals {
  subTotal: number;
  discountTotal: number;
  cgst: number;
  sgst: number;
  igst: number;
  taxTotal: number;
  roundOff: number;
  grandTotal: number;
}

export function previewTotals(lines: PreviewLine[], interState = false): PreviewTotals {
  let subTotal = 0;
  let discountTotal = 0;
  let cgst = 0;
  let sgst = 0;
  let igst = 0;

  for (const l of lines) {
    const taxable = Math.max(0, l.unitPrice * l.quantity - l.discount);
    subTotal += taxable;
    discountTotal += l.discount;
    if (interState) {
      igst += Math.round((taxable * l.gstRate) / 100);
    } else {
      cgst += Math.round((taxable * (l.gstRate / 2)) / 100);
      sgst += Math.round((taxable * (l.gstRate / 2)) / 100);
    }
  }

  const taxTotal = cgst + sgst + igst;
  const preRound = subTotal + taxTotal;
  const grandTotal = Math.round(preRound / 100) * 100;
  const roundOff = grandTotal - preRound;

  return { subTotal, discountTotal, cgst, sgst, igst, taxTotal, roundOff, grandTotal };
}

File: internal/service/gst.go
package service

import "autoshop/internal/pkg/money"

// GST engine — pure functions, no database. This is the tax heart of billing and
// is exhaustively unit-tested. India GST rules implemented:
//
//   - Tax is charged on the taxable value = (unit price × qty) − line discount.
//   - Intra-state sale: tax splits equally into CGST + SGST (rate/2 each).
//   - Inter-state sale: tax charged wholly as IGST (full rate).
//   - Each line's tax is rounded to the nearest paise (in money.Percent).
//   - The final payable is rounded to the nearest rupee; the difference is
//     recorded as RoundOff (standard on Indian invoices).

// GSTLineInput is one billing line fed to the engine.
type GSTLineInput struct {
	UnitPrice money.Money
	Quantity  int
	Discount  money.Money // absolute, on the whole line
	GSTRate   float64     // percent, e.g. 18
}

// GSTLineResult is the computed tax breakdown for one line.
type GSTLineResult struct {
	TaxableValue money.Money
	CGST         money.Money
	SGST         money.Money
	IGST         money.Money
	LineTotal    money.Money
}

// GSTResult is the fully computed invoice tax summary.
type GSTResult struct {
	Lines         []GSTLineResult
	SubTotal      money.Money // sum of taxable values
	DiscountTotal money.Money // sum of line discounts
	CGST          money.Money
	SGST          money.Money
	IGST          money.Money
	TaxTotal      money.Money
	RoundOff      money.Money // adjustment to reach a whole-rupee grand total
	GrandTotal    money.Money // final payable (rounded to nearest rupee)
}

// CalculateGST computes the tax breakdown for a set of lines. interState selects
// IGST (true) vs CGST/SGST (false).
func CalculateGST(lines []GSTLineInput, interState bool) GSTResult {
	res := GSTResult{Lines: make([]GSTLineResult, 0, len(lines))}

	for _, in := range lines {
		gross := in.UnitPrice.MulQty(in.Quantity)
		taxable := gross.Sub(in.Discount)
		if taxable.IsNegative() {
			taxable = money.Zero
		}

		var lr GSTLineResult
		lr.TaxableValue = taxable
		if interState {
			lr.IGST = taxable.Percent(in.GSTRate)
		} else {
			half := in.GSTRate / 2
			lr.CGST = taxable.Percent(half)
			lr.SGST = taxable.Percent(half)
		}
		lr.LineTotal = taxable.Add(lr.CGST).Add(lr.SGST).Add(lr.IGST)

		res.Lines = append(res.Lines, lr)
		res.SubTotal = res.SubTotal.Add(taxable)
		res.DiscountTotal = res.DiscountTotal.Add(in.Discount)
		res.CGST = res.CGST.Add(lr.CGST)
		res.SGST = res.SGST.Add(lr.SGST)
		res.IGST = res.IGST.Add(lr.IGST)
	}

	res.TaxTotal = res.CGST.Add(res.SGST).Add(res.IGST)
	preRound := res.SubTotal.Add(res.TaxTotal)
	rounded := roundToRupee(preRound)
	res.RoundOff = rounded.Sub(preRound)
	res.GrandTotal = rounded
	return res
}

// roundToRupee rounds an amount to the nearest whole rupee (100 paise).
func roundToRupee(m money.Money) money.Money {
	p := m.Paise()
	neg := p < 0
	if neg {
		p = -p
	}
	r := ((p + 50) / 100) * 100
	if neg {
		r = -r
	}
	return money.FromPaise(r)
}
